### ISSUE TITLE
Nested truncations are not serializable

### ISSUE THREAD

**Issue by yonjah** [sequelize/sequelize#9770]
## What are you doing?
When using nested transaction, internal transactions queries run without considering the internal savepoint state -
Example code -
```sql
CREATE TABLE `testTrans` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `n1` bigint(20),
  `n2` bigint(20),
  `updatedAt` timestamp NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
  `createdAt` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`id`)
);
```
```js
const Sequelize = require('sequelize');
const db  = new Sequelize('test', 'root', null, {dialect: 'mysql'});
const Table = db.define('testTrans', {n1: {type: Sequelize.INTEGER}, n2: {type: Sequelize.INTEGER}});

db.transaction((t1) => {
  const arr = [];
  for (let i = 1; i <= 3; i += 1) {
    arr.push({n1: i, n2: i * 10});
  }

  return Sequelize.Promise.map(arr, (item) =>
    Table.create(item, {transaction: t1})
      .then((inst) => db.transaction({transaction: t1}, (t2) =>
        Table.create({n1: inst.n1, n2: inst.n2 + 1}, {transaction: t2})
          .then(() => {
            if (inst.n1 === 2) {
              throw new Error('internal Error');
            }
          })
          .then(() => Table.create({n1: inst.n1, n2: inst.n2 + 2}, {transaction: t2}))
          .then(() => Table.create({n1: inst.n1, n2: inst.n2 + 3}, {transaction: t2}))
      )).catch((e) => {
        // handle error but keep running other internal transactions
      })
  );
});
```
## What do you expect to happen?
The code will create a few instances but will throw in internal transaction if n1 is 2
so all instances should be created other than (2,22),(2,23) and (2,21) which should be removed when rolling back

I would expect queries to run something like -
```log
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): START TRANSACTION;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,10,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,2,20,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,30,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-1`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,11,'2018-08-07 08:49:26','2018-08-07 07:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,12,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,13,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-2`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,2,21,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): ROLLBACK TO SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-2`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-3`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,31,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,32,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,33,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): COMMIT;
```

and that the data will finally look something like this

```console
| id | n1   | n2   | updatedAt           | createdAt           |
+----+------+------+---------------------+---------------------+
|  1 |    1 |   10 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  2 |    2 |   20 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  3 |    3 |   30 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  4 |    1 |  11 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  5 |    1 |   12 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  6 |    1 |   13 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  7 |    3 |   31 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  8 |    3 |   32 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
| 9 |    3 |   33 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
+----+------+------+---------------------+---------------------+
```

## What is actually happening?
if we look at the execution log we see the savepoints queries are running concurrently with no specific 
order
```log
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): START TRANSACTION;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,10,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,2,20,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,30,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-1`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-2`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-3`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,11,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,2,21,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,31,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,12,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): ROLLBACK TO SAVEPOINT `56c3b09a-9af8-411a-839a-f8c47035506d-sp-2`;
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,32,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,1,13,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): INSERT INTO `testTrans` (`id`,`n1`,`n2`,`createdAt`,`updatedAt`) VALUES (DEFAULT,3,33,'2018-08-07 08:49:26','2018-08-07 08:49:26');
Executing (56c3b09a-9af8-411a-839a-f8c47035506d): COMMIT;
```
Since there is no specific order in which the queries are run the final data is missing a few records
```console
| id | n1   | n2   | updatedAt           | createdAt           |
+----+------+------+---------------------+---------------------+
|  1 |    1 |   10 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  2 |    2 |   20 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  3 |    3 |   30 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  8 |    3 |   32 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
|  9 |    1 |   13 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
| 10 |    3 |   33 | 2018-08-07 08:49:26 | 2018-08-07 08:49:26 |
+----+------+------+---------------------+---------------------+
```

__Dialect:__ mysql
__Dialect version: 1.5.3
__Database version:__ 10.1.34-MariaDB MariaDB Server
__Sequelize version:__ 4.37.10
__Tested with latest release:__ No

**Note :** _Your issue may be ignored OR closed by maintainers if it's not tested against latest version OR does not follow issue template._

**Comment by yonjah** [comment]
> Btw, I think sequelize can detect that kind of situations: for example, if transaction already has a pending child transaction, opening a new child should wait for child committed.

@Strate I don't think it can I think it has to, that's the whole point of this bug :-)
Sure you can solve all this stuff in user code making sure you don't run anything in parallel, but it's not going to be easy especially if your code is more complex than this very basic example.

The way it is now nested transactions and the savepoints being created have more chance of causing damage than being useful in the case of a rollback

**Comment by yonjah** [comment]
@sushantdhiman Any more information needed for this issue ?

**Comment by sushantdhiman** [comment]
Interesting issue, currently Sequelize is firing whole tree of nested transactions in one go that indeed makes partial rollbacks cumbersome.

Do you have any example how other frameworks are handling nested transactions, in terms of ordering. I will accept this change but transaction implementation may needs to be changed.

I understand this "may" result in slower transactions for deeply nested case as now we will wait for old child transaction to finish before executing next one, but this is the correct behaviour

**Comment by yonjah** [comment]
@sushantdhiman I don't really know how other frameworks are handling this issue but I'll try to check.

I don't really see the performance case here.
If you actually care about atomic operations you'll want the queries you have inside a transaction / save point to run in the specific order without any leaks from other queries.
So assuming you really needed the child transaction you expect the underline run to be sequential.
If you don't really need the set of operation to be atomic and the performance hit is too big you just need to use the parent transaction and you should not create a child transactions.

The main issue I see with this change is that if we actually wait for a child transaction to finish before running other queries there is a chance we lock the transaction completely out if the developer accidentally uses the parent transaction instead of the child.
```js
db.transaction((parent) => {
  return Sequelize.Promise.map(arr, (item) =>
    Table.create(item, {transaction: parent})
      .then((inst) => db.transaction({transaction: parent}, (child) =>
        Table.create({n1: inst.n1, n2: inst.n2 + 1}, {transaction: parent}) //oops
      ))
  );
});
```
This is surely a bug in user code but since until now there was no such enforcement and any way child transactions didn't change the actual sequence of queries it didn't have any actual affect. So there is a chance there are apps in the wild that will suffer from this issue.

Second if you develop an app and accidentally create such a lock it can be really hard to find and debug what halting your app. 
Unfortunately there are semi legitimate ways to use the parent while a child is still active -
```js
db.transaction((parent) => {
  return Sequelize.Promise.all([
    Table.findOrCreate({item: 1}, {transaction: parent}),
    Table.findOrCreate({item: 2}, {transaction: parent}),
    OtherTable.findAll({transaction: parent})
  ]);
});
```
In this case `findOrCreate` will create internal transactions we don't really have control on but the `findAll` will still use the parent transaction as is. this code will not lock the transaction since the child transactions don't care about the calls that are related to the parent.

The options I see to help mitigate this issue are -
1. Add ttl for query/transaction execution that will throw if they wait for too long
2. Have strict mode option that will warn/throw if a query is sent to the parent while a child is still active
3. Have some easy way for getting info about what transactions / queries are currently waiting (with execution stack) so if your app halt you can inspect it and find the cause

I don't like the first option that much cause ttls has a tendency to hit you in production at the worst time possible. 
The second might still be annoying / unusable if you have a lot of cases where it's not actually locking. The third option is nice to have for debugging but will not protect you in anyway if you actually have such a lock in your code.

I'm not sure what the best option though and I'm probably over thinking this whole issue

**Comment by yonjah** [comment]
Update I checked Knex and it seem to suffer from the same issue -
https://github.com/tgriesser/knex/issues/2213

I also checked Waterline but if I understand the [docs](https://github.com/balderdashy/sails-docs/blob/master/reference/waterline/datastores/transaction.md) correctly their concept of transactions seem to give you back a connection instance to lock queries to, so I don't think they support nesting or savepoints

**Comment by sushantdhiman** [comment]
I was looking into Spring framework, looks like they support both `ttl` and have strict ordering. We can look into introducing these concepts.

If we talk about `ttl`, each transaction currently run for timeout specified by database engine so it's database dependent, not standardised but I guess safer, because we have no actual ability to cancel statements.

Strict ordering is our main goal, we can introduce strict mode for transactions, under this mode transaction and savepoints will be ordered. This will allow us to introduce this change as non breaking

**Comment by yonjah** [comment]
Having a strict mode sounds good to me.

I'm not 100% sure about the built in ttl though. Not really sure how it's going to work, but wont the connection just be closed ? could make bugs hard to detect and debug

**Comment by sushantdhiman** [comment]
> I'm not 100% sure about the built in ttl though. Not really sure how it's going to work, but wont the connection just be closed ? could make bugs hard to detect and debug

Connection is never closed after transaction, merely returned to pool so it can be used by other calls. As we lack cancelling statement support let's not implement `ttl` for now.

In future `ttl` should work something like this, transaction reached maximum allotted time, we can cancel statement and throw timeout. That error can bubble up in transaction chain so parent transaction can rollback itself without affecting other siblings.

For now we just need to implement strict ordering

**Comment by Strate** [comment]
I would disagree to have a strict mode, to have a serial nested transactions. It should be one and only possible behaviour, because any other could possible introduce bugs.

**Comment by yonjah** [comment]
@Strate I agree that this is the only correct behavior, but I think @sushantdhiman suggesting to add it for now as a mode so we shouldn't worry about breaking changes and can introduce the fix in current version.

@sushantdhiman not sure I understand the cancel statement issue. I don't know the internals of transactions that much and you might know of a better way to implement it, but I would have thought that If sequelize is taking care of ordering statements would not be sent to the db unless they are in the correct order. So it's not the db that going to halt a specific statement but Sequelize waiting without sending any statements to the db.
